Update leave_policy/parallel phase-related RFCs (#2130)
oneapi-src/oneTBB: rfcs/experimental/global_control_fast_leave/README.md

This proposal introduces a new `global_control` parameter called `leave_policy` that, when set to "fast",
would override the default behavior for arenas that are initialized with `leave_policy::automatic`.
Setting the parameter would not affect arenas that are already initialized or initialized with an
explicit `leave_policy`. This follows the decision made in the
[Completely disable new behavior](../parallel_phase_for_task_arena/README.md#completely-disable-new-behavior)
section of the parallel phase RFC to keep the arena leave policy static for the lifetime of the arena,

> since there are no use cases for dynamic transition between the states yet. 

Reviewer: ```suggestion
since we are not aware of practical use cases for dynamic transition between the states. 
```

Reviewer: There may be use cases already, we just don't know about them.

| `task_arena::leave_policy::automatic` (default) | Workers follow the default system-specific policy (may spin before leaving) |
| `task_arena::leave_policy::fast` | Workers leave immediately (fast leave enabled) |

When multiple `global_control` objects exist for `leave_policy`, their logical disjunction is
used (consistent with the `terminate_on_exception` parameter). For `leave_policy`, the request for
`fast` is the restrictive one, since `automatic` leaves the choice up to the library, so it wins over
the absence of such a request. If more leave policies appear, a new strategy of choosing the
`leave_policy` can be implemented. At this point, it means that if any

> `global_control(leave_policy, task_arena::leave_policy::fast)` is active, fast leave is enabled globally.

Reviewer 2: I suggest not to call it "disjunction" in the documentation, especially the public one.

We should describe the behavior in the way that a) gives certain clear guarantees, and b) allows to extend/modify that in the future, ideally without breaking these guarantees. The current guarantee, I think is stated in the last sentence - so let's make it clear and upfront, and then clarify as necessary.

At the same time, I think this stated guarantee will be hard to keep if another policy is ever added - because of "any". Most likely in that case we will have to break the behavior.

Author: Updated.

| `task_arena::leave_policy::fast` | Workers leave immediately (fast leave enabled) |

> When multiple `global_control` objects exist for `leave_policy`, their logical disjunction would be

Reviewer: ```suggestion
When multiple `global_control` objects exist for `leave_policy`, their logical disjunction is
```

used (consistent with the `terminate_on_exception` parameter). For `leave_policy`, the request for
`fast` is the restrictive one, since it asks the library to give up threads sooner, so it wins over

> the absence of such a request. As more leave policies appear, new strategy of choosing the next

Reviewer: ```suggestion
`fast` is the restrictive one, since `automatic` leaves the choice up to the library, so it wins over
the absence of such a request. If more leave policies appear, a new strategy of choosing the
```

`leave_policy` can be implemented. At this point, it means that if any

> `global_control(leave_policy, task_arena::leave_policy::fast)` is active, fast leave would be enabled globally.

Reviewer: ```suggestion
`global_control(leave_policy, task_arena::leave_policy::fast)` is active, fast leave is enabled globally.
```
